crates/libs/windows/src/core/pcwstr.rs

```diff
@@ -45,7 +45,7 @@ unsafe impl Abi for PCWSTR {
 #[cfg(feature = "alloc")]
 impl<'a> IntoParam<'a, PCWSTR> for &str {
     fn into_param(self) -> Param<'a, PCWSTR> {
-        Param::Boxed(PCWSTR(heap_string(&self.encode_utf16().collect::<alloc::vec::Vec<u16>>())))
+        Param::Boxed(PCWSTR(unsafe { string_from_iter(self.encode_utf16(), self.len()) }))
```

Comment:
Will `self.len() >= self.encode_utf16().count()` always hold true?  Intuitively it seems like it should, a 3-byte UTF-8 sequence representing two code points could be represented by either two `u16` if they combine, or three `u16` if they map 1:1.  Worst case is that it  over-allocates.

Reply:
I'd been banging my head at this for a while now, too. It seems to hold true, so long as `encode_utf16()` doesn't do anything crazy, like decomposing code points.

The rationale I came up with is this: Each Unicode code point can be encoded as no more than two UTF-16 code units. For the inequality to hold, we can thus ignore all UTF-8 sequences that consist of two or more code units. That leaves the set of code points that require a single UTF-8 code unit. Those are the 128 ASCII characters, which are all encoded using a single code unit in UTF-16 as well.

In summary:

* UTF-8 sequences of length 1 map to UTF-16 sequences of length 1
* UTF-8 sequences of length 2 or more map the UTF-16 sequences of length 2 (or less)

While I couldn't find any formal documentation on what `encode_utf16()` does and doesn't do, the example listed under [`encode_utf16()`](https://doc.rust-lang.org/std/primitive.str.html#method.encode_utf16) at least seems to confirm the reasoning above: `assert!(utf16_len <= utf8_len);`.

Still a bit uneasy about all of this. Personally, I'd prefer a `string_from_iter()` that takes an iterator only, wastes some cycles on `count()` (ultimately doing the de-/encoding twice), and drops the `unsafe` along the way.

Thoughts?

Reply:
Yes it will given three assumptions:

1. the encoding of a Unicode code point in UTF-8 characters will always be higher or equal to the number of UTF-16 characters. UTF-8 is a variable-width character encoding that takes [four possible forms:](https://howardhinnant.github.io/utf_summary.html)

1 UTF-8 Character => 1 UTF-16 Character
2 UTF-8 Characters => 1 UTF-16 Character
3 UTF-8 Characters => 1 UTF-16 Character
4 UTF-8 Characters => 2 UTF-16 Characters

2. No (de)normalization takes place / no decomposing or composing character transformations (e.g. A + ¨: LATIN CAPITAL LETTER A (U+0041) + COMBINING DIAERESIS (U+0308)). There is no indication in the documentation of the Rust standard library that any of such normal or transformations takes place and looking at the source code only validates that assumption.

3. Input is valid UTF-8. We might have to look into edge-cases around OsStr, since it consists of [WTF-8 strings](https://simonsapin.github.io/wtf-8/).

Reply:
For the sake of clarity here's an exhaustive table of UTF-8 code units to UTF-16 code units.

|UTF-8|UTF-16|
|---|---|
|1|1|
|2|1|
|3|1|
|4|2|

`encode_utf16` does no magic. It's a straightforward re-encoding of UTF-8 to UTF-16.

Reply:
> Personally, I'd prefer a `string_from_iter()` that takes an iterator only, wastes some cycles on `count()` (ultimately doing the de-/encoding twice), and drops the `unsafe` along the way.

Is there a guarantee that an iterator can be iterated twice?

I doubt there's any de/normalization taking place in `encode_utf16()`  I'd be willing to bet dozens of dollars on it!  Survey says...  There's no denormalization, just bit twiddling as expected:  https://doc.rust-lang.org/stable/src/core/char/methods.rs.html#1706

Also, the Rust documentation for `encode_utf16()` seems to indicate that the utf-16 `u16` count is strictly less than the utf-8 byte count: 
 https://doc.rust-lang.org/stable/src/core/str/mod.rs.html#998

Reply:
> Also, the Rust documentation for encode_utf16() seems to indicate that the utf-16 u16 count is strictly less than the utf-8 byte count:
https://doc.rust-lang.org/stable/src/core/str/mod.rs.html#998

Nice find. Even though it's just an example I'd say it was intended as a guarantee.

The way I see it there are three ways in which we can resolve this:

1. For each core::ptr::write we do, we assert that we are within bounds first (panic behaviour)
2. We run [.take(len + 1)](https://doc.rust-lang.org/core/iter/trait.Iterator.html#method.take) on the iterator, which truncates any theoretical superfluous characters.
3. We rethink our approach with iterators.

To be honest I'm not very happy about this function to begin with. I designed it the way I did in order to remain "backwards compatible" to `heap_string`. We are currently generic over T with an iterator we can only iterate once. 

Why do we need to be generic over T? The only strings that the Windows API uses are UTF-8 and UTF-16 and some day there might be a use case for UTF-32. These have an alignment of 1, 2 and 4 respectively. As we are generic currently, we assert that the alignment of the output buffer matches the alignment of T. However, given that [`HeapAlloc`](https://docs.microsoft.com/en-us/windows/win32/api/heapapi/nf-heapapi-heapalloc) has an alignment of 8+ and that we only ever use 4 as maximum alignment, that check is fairly redundant.

As it stands, this API has the following inputs: `&OsStr` and `&str`. And the following outputs: `*const u8` and `*const u16`. We could create two functions instead: `heap_string_narrow`, `heap_string_wide` for `*const u8` and `*const u16`, respectively. This would allow us to iterate over the string any number of times and ensure they are within bounds.

As I was writing this I actually have an even better idea, hold on let me implement it.